utils/utf8.cc

```diff
@@ -56,22 +56,30 @@ namespace utils {
 
 namespace utf8 {
 
-// 3x faster than boost utf_to_utf
-static inline std::optional<size_t> validate_naive(const uint8_t *data, size_t len) {
-    size_t pos = 0;
-
-    while (len) {
-        size_t bytes;
-        const uint8_t byte1 = data[0];
+struct codepoint_status {
+    size_t bytes_validated;
+    bool error;
+    uint8_t more_bytes_needed;
+};
 
+static
+codepoint_status
+inline
+evaluate_codepoint(const uint8_t* data, size_t len) {
+    auto byte1 = data[0];
+    static uint8_t len_from_first_nibble[16] = { 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 3, 4 };
+    auto codepoint_len = len_from_first_nibble[byte1 >> 4];
+    if (codepoint_len > len) {
+	return codepoint_status{.more_bytes_needed = uint8_t(codepoint_len - len)};
```

Comment:
I think there's a bug here. Even if codepoint_len > len, you still have "len" to check. I think you need to check that all the bytes that you do have after the first one are in the range 0x80 and 0xBF, and if not, set "error".
In fact, that's *all* you need to do - it seems to me most of the various cases below are not really needed (more on this below), and moreover, the whole idea of remembering the end of the buffer is not needed. See more on that too.

Reply:
Ah, I think it's protecting against overlong sequences.

Reply:
Oh, I *think* maybe I understand now... If you have a 4-byte sequence, and only got 3, you don't check these 3 at all. You just collect them to later, and work on the 4 together, later. Is that true? If so, than the fact the code returns more_bytes_needed immediately, without checking anything, is correct, and there is no bug.

Reply:
Yes, checking is performed when we have a full codepoint.